src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +30,109 @@ public static class Renamer
+            }
+
+            if (shouldMoveNamespace)
+            {
+                solution = await SyncNamespaceToDocumentFoldersAsync(document.Id, solution, cancellationToken).ConfigureAwait(false);
+            }
+
+            return new RenameDocumentResult(solution, renameSymbolSucceeded: true);
+        }
+
+        private static async Task<Solution> RenameTypeToMatchDocumentNameAsync(DocumentId documentId, Solution solution, string newName, string originalName, CancellationToken cancellationToken)
+        {
+            var newNameWithoutExtension = Path.GetFileNameWithoutExtension(newName);
+            var oldNameWithoutExtension = Path.GetFileNameWithoutExtension(originalName);
+
+            var document = solution.GetDocument(documentId);
+
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            var typeDeclarations = syntaxRoot.DescendantNodesAndSelf().Where(syntaxFacts.IsTypeDeclaration);
+            var displayNames = typeDeclarations.Select(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None));
+
+            var matchingTypeDeclaration = typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(oldNameWithoutExtension, StringComparison.OrdinalIgnoreCase));
+            if (matchingTypeDeclaration is object)
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                solution = await RenameSymbolAsync(solution, symbol, newNameWithoutExtension, solution.Options, cancellationToken).ConfigureAwait(false);
+            }
+
+            return solution;
+        }
+
+        private static async Task<Solution> SyncNamespaceToDocumentFoldersAsync(DocumentId documentId, Solution solution, CancellationToken cancellationToken)
+        {
+            var document = solution.GetDocument(documentId);
+            var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var targetNamespace = BuildNamespaceFromFolders(document.Folders, syntaxFacts);
+
+            var syntaxRoot = await document.GetSyntaxRootAsync().ConfigureAwait(false);
+
+            var namespaceDeclarations = syntaxRoot.DescendantNodes(descendIntoChildren: n => !syntaxFacts.IsDeclaration(n)).Where(n => !syntaxFacts.IsTypeDeclaration(n));
+
+            foreach (var namespaceDeclaration in namespaceDeclarations)
+            {
+                solution = await changeNamespaceService.ChangeNamespaceAsync(
+                    document,
+                    namespaceDeclaration,
```

Comment:
i don't know how this could work for namespaces 2-N.  Once you change once namespace, you'll have a new document, and the old namespace decl you're referring to has no rlation to it.